src/metabase/query_processor/middleware/wrap_value_literals.clj

```diff
@@ -43,12 +49,27 @@
      (when (types/temporal-field? field-info)
        {:unit :default}))))
 
+(defn- str-id-field->type-info
+  "Return _type info_ for a field with `field-name`, presumably from source query model.
+
+  Idea is, that if field has a string id, it may come from a model. If it _comes_ from a model, the model has to be
+  a source query. Hence the query with source model contains `:source-metadata` those could be mapped to `field-name`.
```

Comment:
What if it doesn't come from the source?

Reply:
I assume your point is it may come from joins. Thanks for mentioning that. I've modified the code so it is respected -- appropriate metadatas are chosen first and searched for the entry corresponding to a field afterwards.

Reply:
I believe there should not be clashes with expressions. Experimentally, if I use query as from attached screenshot, names of expression and field are deduplicated in query and also in source metadata. The source model contains expression and field with name `status`.

<img width="1484" alt="Screenshot 2024-08-12 at 18 14 14" src="https://github.com/user-attachments/assets/6fa1988a-e624-45d8-bf96-9ea540913856">